CUDADataFormats/HcalDigi/interface/DigiCollection.h

```diff
@@ -0,0 +1,187 @@
+#ifndef CUDADataFormats_HcalDigi_interface_DigiCollection_h
+#define CUDADataFormats_HcalDigi_interface_DigiCollection_h
+
+#include "CUDADataFormats/CaloCommon/interface/Common.h"
+
+namespace hcal {
+
+  struct Flavor01 {
+    using adc_type = uint8_t;
+    using tdc_type = uint8_t;
+    using soibit_type = uint8_t;
+
+    static constexpr int WORDS_PER_SAMPLE = 1;
+    static constexpr int HEADER_WORDS = 1;
+
+    static constexpr adc_type adc(uint16_t const* const sample_start) { return (*sample_start & 0xff); }
+
+    static constexpr tdc_type tdc(uint16_t const* const sample_start) { return (*sample_start >> 8) & 0x3f; }
+
+    static constexpr soibit_type soibit(uint16_t const* const sample_start) { return (*sample_start >> 14) & 0x1; }
+  };
+
+  struct Flavor2 {
+    static constexpr int WORDS_PER_SAMPLE = 2;
+    static constexpr int HEADER_WORDS = 1;
+  };
+
+  struct Flavor3 {
+    using adc_type = uint8_t;
+    using tdc_type = uint8_t;
+    using soibit_type = uint8_t;
+
+    static constexpr int WORDS_PER_SAMPLE = 1;
+    static constexpr int HEADER_WORDS = 1;
+
+    static constexpr adc_type adc(uint16_t const* const sample_start) { return (*sample_start & 0xff); }
+
+    static constexpr tdc_type tdc(uint16_t const* const sample_start) { return ((*sample_start >> 8) & 0x3); }
+
+    static constexpr soibit_type soibit(uint16_t const* const sample_start) { return ((*sample_start >> 14) & 0x1); }
+
+    static constexpr uint8_t capid(uint16_t const* const sample_start) { return ((*sample_start >> 10) & 0x3); }
+  };
+
+  struct Flavor4 {
+    static constexpr int WORDS_PER_SAMPLE = 1;
+    static constexpr int HEADER_WORDS = 1;
+  };
+
+  struct Flavor5 {
+    using adc_type = uint8_t;
+
+    static constexpr float WORDS_PER_SAMPLE = 0.5;
+    static constexpr int SAMPLES_PER_WORD = 2;
+    static constexpr int HEADER_WORDS = 1;
+
+    static constexpr adc_type adc(uint16_t const* const sample_start, uint8_t const shifter) {
+      return ((*sample_start >> shifter * 8) & 0x7f);
+    }
+  };
+
+  template <typename Flavor>
+  constexpr uint8_t capid_for_sample(uint16_t const* const dfstart, uint32_t const sample) {
```

Comment:
I admit I don't know what this code is supposed to do - but I don't understand the comment ?

Reply:
@fwyzard (it's a minor point but) I don't understand why at lines 25-27 above the types for adc, tdc and soibit are named, and then used coherently afterwards in the code, but not the one for capid

Reply:
there is no particular reason

Reply:
Ah, I see.

Do we actually expect the types to be different for the different cases ?
Otherwise, wouldn't it be simpler to just drop all typdefs and use `uint8_t` everywhere ?

Reply:
i do not remember the logic, but yeah at this point here we can simplify...